agent/tcs/client/client.go

```diff
@@ -149,30 +147,85 @@ func (cs *clientServer) publishMetrics() {
 
 // publishMetricsOnce is invoked by the ticker to periodically publish metrics to backend.
 func (cs *clientServer) publishMetricsOnce() {
-	metadata, taskMetrics, err := cs.statsEngine.GetInstanceMetrics()
+	// Get the list of objects to send to backend.
+	requests, err := cs.metricsToPublishMetricRequests()
 	if err != nil {
 		log.Warn("Error getting instance metrics", "err", err)
-		return
 	}
 
-	if *metadata.Idle {
-		// Idle instance, send message and return.
-		cs.MakeRequest(ecstcs.NewPublishMetricsRequest(metadata, taskMetrics))
-		return
+	// Make the publish metrics request to the backend.
+	for _, request := range requests {
+		cs.MakeRequest(request)
 	}
+}
 
+// metricsToPublishMetricRequests gets task metrics and converts them to a list of PublishMetricRequest
+// objects.
+func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetricsRequest, error) {
+	metadata, taskMetrics, err := cs.statsEngine.GetInstanceMetrics()
+	if err != nil {
+		return nil, err
+	}
+
+	var requests []*ecstcs.PublishMetricsRequest
+	if *metadata.Idle {
+		fin := true
+		metadata.Fin = &fin
+		// Idle instance, we have only one request to send to backend.
+		requests = append(requests, ecstcs.NewPublishMetricsRequest(metadata, taskMetrics))
+		return requests, nil
+	}
 	var messageTaskMetrics []*ecstcs.TaskMetric
-	for i := range taskMetrics {
-		messageTaskMetrics = append(messageTaskMetrics, taskMetrics[i])
+	numTasks := len(taskMetrics)
+
+	for i, taskMetric := range taskMetrics {
+		messageTaskMetrics = append(messageTaskMetrics, taskMetric)
+		var fin bool
+		if (i + 1) == numTasks {
+			// If this is the last task to send, set fin to true
+			fin = true
+		} else {
+			fin = false
+		}
+		requestMetadata := fromMetricsMetadata(metadata, fin)
 		if (i+1)%tasksInMessage == 0 {
 			// Construct payload with tasksInMessage number of task metrics and send to backend.
-			cs.MakeRequest(ecstcs.NewPublishMetricsRequest(metadata, messageTaskMetrics))
+			requests = append(requests, ecstcs.NewPublishMetricsRequest(requestMetadata, copyTaskMetrics(messageTaskMetrics)))
 			messageTaskMetrics = messageTaskMetrics[:0]
 		}
 	}
 
 	if len(messageTaskMetrics) > 0 {
-		// Send remaining task metrics to backend.
-		cs.MakeRequest(ecstcs.NewPublishMetricsRequest(metadata, messageTaskMetrics))
+		fin := true
+		requestMetadata := fromMetricsMetadata(metadata, fin)
```

Comment:
fin is meant to indicate that this is the final message in the collection of metric message objects in the model. Using the variable here was meant to makes it clear what the 'true' value meant. A comment would also work here. I will change it as I think the comment would make it more clear.